File: tst/Kroki.Tests/Output/Fringe.cs
using System;
using System.Collections.Generic;
using System.Text;
using Kroki.Runtime;

namespace Kroki.Example
{
    public delegate int TStartFunc(int owner);
    public delegate int TStartFuncEx(int owner, bool pSkinInterface);
    public delegate void TStartAction(int owner, float hello);
    public delegate bool TWTSRegisterSessionNotification(int Wnd, uint dwFlags);
    public class characterChoice1 : List<string>
    {
    }

    public class characterChoice2 : List<string>
    {
    }

    public enum TSpeed
    {
        spVerySlow,
        spSlow,
        spAverage,
        spFast,
        spVeryFast
    }

    public class TPossibleSpeeds : List<TSpeed>
    {
    }

    public struct ptrToInt2
    {
    }

    public class signedQword
    {
        public ushort value;
        public int signum;
    }

    public struct signedQwordPacked
    {
        public ushort value;
        public int signum;
    }

    public class TA : object
    {
        private int GetA()
        {
            return -1;
        }

        private void SetA(int AValue)
        {
            Console.WriteLine(AValue);
        }

        public void Create()
        {
            Console.WriteLine("Class constructor TA");
        }

        public void Destroy()
        {
            Console.WriteLine("Class destructor TA");
        }

        public int A { get; set; }
    }

    public interface IMyDelegate
    {
        void DoThis(int value);
    }

    public class TMyClass : IMyDelegate
    {
        public void DoThis(int value)
        {
            string Str = default;
            Console.WriteLine("Success!!! Type <enter> to continue ", value);
            Console.ReadLine(Str);
        }
    }

    public static class fringe
    {
        public static bool IsFilterConnected()
        {
            return true;
        }

        public static void Comp(bool FGrabsSamples)
        {
            if (FGrabsSamples)
            {
                if (!IsFilterConnected())
                {
                    throw Exception.Create("Sample grabber not connected");
                }
            }
        }

        public static void DoFor(List<String> Items, long width, long height)
        {
            string menuItem = default;
            int I = default;
            foreach (var menuItem in Items)
            {
                Console.WriteLine(" (" + menuItem + ")");
            }

            for (I = 0; I <= width * height - 1; I++)
            {
                Console.WriteLine(I);
            }
        }

        public static bool CtrlDown()
        {
            int State = 453;
            Console.WriteLine(State);
            return ((State && 128) != 0);
        }

        public static void DrawPart()
        {
            float part = 4.223;
            Console.Write(part);
        }

        public static int DoTry(List<String> FLog, string logmsg)
        {
            int res = default;
            try
            {
                if (FLog != null)
                {
                    FLog.Add(logmsg);
                }
            }
            catch
            {
                FLog = null;
            }

            try
            {
                res = 42;
            }
            catch (Exception E)
            {
                res = -1;
            }

            return res;
        }

        public static void DoAssmbl()
        {
        }

        public delegate hResult TSHGetImageList(int iImageList, TGUID riid);
        public static string ImageListExtraLarge()
        {
            THandle hInstShell32 = default;
            TSHGetImageList SHGetImageList = default;
            const int SHIL_LARGE = 0;
            const int SHIL_SMALL = 1;
            const int SHIL_EXTRALARGE = 2;
            const string IID_IImageList = "{46EB5926-582E-4017-9FDF-E8998DAA0950}";
            hInstShell32 = LoadLibrary("Shell32.dll");
            if (hInstShell32 != 0)
            {
                try
                {
                    Console.WriteLine(hInstShell32, SHIL_LARGE, SHIL_EXTRALARGE, SHIL_SMALL, SHGetImageList(0, IID_IImageList));
                }
                finally
                {
                    FreeLibrary(hInstShell32);
                }
            }

            return "?";
        }

        public delegate bool TWTSRegisterSessionNotification(int Wnd, uint dwFlags);
        public static bool RegisterSessionNotification(int Wnd, uint dwFlags)
        {
            THandle hWTSapi32dll = default;
            TWTSRegisterSessionNotification WTSRegisterSessionNotification = default;
            hWTSapi32dll = LoadLibrary("Wtsapi32.dll");
            if (hWTSapi32dll > 0)
            {
                try
                {
                    return WTSRegisterSessionNotification(Wnd, dwFlags);
                }
                finally
                {
                    if (hWTSapi32dll > 0)
                    {
                        FreeLibrary(hWTSAPI32DLL);
                    }
                }
            }

            return false;
        }
    }
}

File: tst/Kroki.Tests/Input/Fringe.pas
unit fringe;

{$H+}

interface

uses
  Classes, SysUtils;

type
  TStartFunc   = function(owner: integer): integer;
  TStartFuncEx = function(owner: integer; pSkinInterface : Boolean): integer;
  TStartAction = procedure(owner: integer; hello: Single);
  TWTSRegisterSessionNotification = function(Wnd: integer; dwFlags: DWORD): Boolean; stdcall;
  characterChoice1 = array[integer] of UCS4char;
  characterChoice2 = array of UCS4char;
  TSpeed = (spVerySlow,spSlow,spAverage,spFast,spVeryFast);
  TPossibleSpeeds = set of TSpeed;
  ptrToInt2 = ^Integer;
  signedQword = record
 	value: qword;
 	signum: -1..1;
  end;
  signedQwordPacked = packed record
	value: qword;
	signum: -1..1;
  end;
  TA = Class(TObject)
  Private
    Function GetA : Integer;
    Procedure SetA(AValue : integer);
  public
    Class Constructor create;
    Class Destructor destroy;
    Property A : Integer Read GetA Write SetA;
  end;
  IMyDelegate = interface
    procedure DoThis (value: integer);
  end;
  TMyClass = class (IMyDelegate)
    procedure DoThis (value: integer);
  end;

implementation

function IsFilterConnected(): Boolean;
begin
  Result := True;
end;

procedure Comp(FGrabsSamples: Boolean);
begin
  if FGrabsSamples then
  begin
    if not IsFilterConnected() then
    begin
      raise Exception.Create('Sample grabber not connected');
    end;
  end;
end;

{Class} Function TA.GetA : Integer;
begin
  Result:=-1;
end;

{Class} Procedure TA.SetA(AValue : integer);
begin
  WriteLn(AValue);
end;

Class Constructor TA.Create;
begin
  Inherited;
  Writeln('Class constructor TA');
end;

Class Destructor TA.Destroy;
begin
  Writeln('Class destructor TA');
end;

procedure DoFor(Items : TStringList; width: LongInt; height: LongInt);
var
  menuItem : String;
  I : integer;
begin
  for menuItem in Items do
    begin
      WriteLn(' (' + menuItem + ')');
    end;
  for I := 0 to width * height - 1 do
    begin
      WriteLn(I);
    end;
end;

function CtrlDown: Boolean;
  var
    State: integer = 453;
  begin
    WriteLn(State);
    Result := ((State and 128) <> 0);
  end;

procedure DrawPart;
  var
    part: Single = 4.223;
  begin
    Write(part);
  end;

function DoTry(FLog : TStringList; logmsg : string): integer;
var
  res : integer;
begin
  try
    if FLog <> nil then FLog.Add(logmsg);
  except
    FLog := nil;
  end;
  try
    res := 42;
  except
    on E:Exception do
    begin
      res := -1;
    end;
  end;
  Result := res;
end;

procedure TMyClass.DoThis(value: integer);
var
  Str: string;
begin
  WriteLn('Success!!! Type <enter> to continue ', value);
  ReadLn(Str);
end;

procedure DoAssmbl();
begin
  asm
      CMP       EDX,EAX
      CMOVG     EAX,EDX
      CMP       ECX,EAX
      CMOVG     EAX,ECX
    mov eax, FS:[4]
    sub eax, 3
  end;
end;

function ImageListExtraLarge: String;
type
  TSHGetImageList = function (iImageList: integer; const riid: TGUID): hResult; stdcall;
var
  hInstShell32: THandle;
  SHGetImageList: TSHGetImageList;
const
  SHIL_LARGE= 0; //32X32
  SHIL_SMALL= 1; //16X16
  SHIL_EXTRALARGE= 2;
  IID_IImageList: TGUID= '{46EB5926-582E-4017-9FDF-E8998DAA0950}';
begin
  hInstShell32:= LoadLibrary('Shell32.dll');
  if hInstShell32<> 0 then
  try
    WriteLn(hInstShell32, SHIL_LARGE, SHIL_EXTRALARGE, SHIL_SMALL, SHGetImageList(0,IID_IImageList));
  finally
    FreeLibrary(hInstShell32);
  end;
  Result:= '?';
end;

function RegisterSessionNotification(Wnd: Integer; dwFlags: Cardinal) : Boolean;
type
  TWTSRegisterSessionNotification = function(Wnd: Integer; dwFlags: DWORD): Boolean; stdcall;
var
  hWTSapi32dll : THandle;
  WTSRegisterSessionNotification : TWTSRegisterSessionNotification;
begin
  hWTSapi32dll := LoadLibrary('Wtsapi32.dll');
  if hWTSapi32dll > 0 then
  begin
    try
      Result:= WTSRegisterSessionNotification(Wnd, dwFlags);
    finally
      if hWTSapi32dll > 0 then
        FreeLibrary(hWTSAPI32DLL);
    end;
  end;
  Result := False;
end;

end.

